### PR TITLE
Introducing list types for feast features retrieval

### PR DESCRIPTION
**Pull request by ariefrahmansyah**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

**Checklist**

- [x] Added unit test, integration, and/or e2e tests
- [x] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: api/pkg/transformer/types/series/series.go [hunk 1]
```diff
@@ -12,10 +12,14 @@ import (
 type Type string
 
 const (
-	String Type = "string"
-	Int    Type = "int"
-	Float  Type = "float"
-	Bool   Type = "bool"
+	String     Type = "string"
```
Comment (tiopramayudi) [inline api/pkg/transformer/types/series/series.go:15]: can you use valueType from common.pb.go instead of this? so we can remove this, and have one single truth of list of types

File: api/models/model_endpoint_alert.go [hunk 2]
```diff
@@ -34,8 +34,8 @@ const (
...
 	errorRateSliExprFormat  = "(100 * sum(rate(revision_request_count{cluster_name=\"%s\",namespace_name=\"%s\",response_code_class!=\"2xx\",revision_name=~\".*%s.*\"}[1m])) / sum(rate(revision_request_count{cluster_name=\"%s\",namespace_name=\"%s\",revision_name=~\".*%s.*\"}[1m])))"
-	cpuSliExprFormat        = "(100 * sum(rate(container_cpu_usage_seconds_total{cluster_name=\"%s\",namespace=\"%s\",pod_name=~\".*%s.*\"}[1m])) / sum(kube_pod_container_resource_requests_cpu_cores{cluster_name=\"%s\",namespace=\"%s\",pod=~\".*%s.*\"}))"
-	memorySliExprFormat     = "(100 * sum(container_memory_usage_bytes{cluster_name=\"%s\",namespace=\"%s\",pod_name=~\".*%s.*\"}) / sum(kube_pod_container_resource_requests_memory_bytes{cluster_name=\"%s\",namespace=\"%s\",pod=~\".*%s.*\"}))"
+	cpuSliExprFormat        = "(100 * sum(rate(container_cpu_usage_seconds_total{cluster_name=\"%s\",namespace=\"%s\",pod=~\".*%s.*\",container!~\"|POD\"}[1m])) / sum(kube_pod_container_resource_requests_cpu_cores{cluster_name=\"%s\",namespace=\"%s\",pod=~\".*%s.*\",container!~\"|POD\"}))"
```
Comment (pradithya) [inline api/models/model_endpoint_alert.go:37]: These seems unrelated changes?

Comment (ariefrahmansyah) [reply api/models/model_endpoint_alert.go:37]: Yeah, it is unrelated changes and I don't know why GitHub shown it here. I think there's an issue with git rebase or github. However, both files in this branch and master has no difference. We uses pod and container label like shown here: `pod=~\".*%s.*\",container!~\"|POD\"`

File: api/pkg/transformer/types/series/series.go [hunk 3]
```diff
@@ -88,6 +97,70 @@ func (s *Series) Get(index int) interface{} {
 	return s.series.Elem(index).Val()
 }
 
+// func (s *Series) Append(values interface{}) {
```
Comment (pradithya) [inline api/pkg/transformer/types/series/series.go:100]: Are these (Append and Concat) being implemented?

File: api/pkg/transformer/types/series/series.go [hunk 4]
```diff
@@ -88,6 +97,70 @@ func (s *Series) Get(index int) interface{} {
...
+// 	s.series.Append(values)
+// }
+
+// func (s *Series) Concat(x Series) *Series {
+// 	concat := s.series.Concat(*x.series)
+// 	return &Series{&concat}
+// }
+
+func (s *Series) Order(reverse bool) *Series {
```
Comment (pradithya) [inline api/pkg/transformer/types/series/series.go:109]: Can you add documentation to all of the series's method? especially for the newly added one

File: api/pkg/transformer/pipeline/testdata/valid_feast_series_transform.yaml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+          how: LEFT
+          onColumns: [driver_id]
+      - tableTransformation:
+          inputTable: result_table
+          outputTable: result_table
+          steps:
+            - updateColumns:
+                - column: driver_feature_3_flatten
+                  expression: result_table.Col('driver_feature_3').Flatten()
+                - column: stddev_driver_score
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/testdata/valid_feast_series_transform.yaml:61]: Can you add tests for other new methods that are introduced in this PR. i.e.: Median, MaxStr, Min, Quantile, Sum, MinStr, Unique?

**Review by pradithya — CHANGES_REQUESTED** [review]

File: api/pkg/transformer/types/series/series.go (hunk 1 above)
Comment (ariefrahmansyah) [reply api/pkg/transformer/types/series/series.go:15]: I put it as out of scope for now as it introduce many changes at many places. Currently we are using string data type to define new series, however our common.pb.go's valueType is int32, hence we need to modify the signature of our merlin's series.New().

File: api/cmd/mock-feast/main.go [hunk 6]
```diff
@@ -30,15 +42,16 @@ func main() {
...
-}
+type mockFeastOnlineServing struct{}
 
 func (m mockFeastOnlineServing) GetFeastServingInfo(ctx context.Context, req *serving.GetFeastServingInfoRequest) (*serving.GetFeastServingInfoResponse, error) {
 	return &serving.GetFeastServingInfoResponse{}, nil
 }
 
 // GetOnlineFeaturesV2 return a valid GetOnlineFeaturesResponse but with all features having GetOnlineFeaturesResponse_OUTSIDE_MAX_AGE
 func (m mockFeastOnlineServing) GetOnlineFeaturesV2(ctx context.Context, req *serving.GetOnlineFeaturesRequestV2) (*serving.GetOnlineFeaturesResponse, error) {
+	log.Infof("GetOnlineFeaturesV2.Request: %+v", req)
```
Comment (pradithya) [inline api/cmd/mock-feast/main.go:53]: Can you set it to debug or remove it? might become very noisy

File: api/pkg/transformer/server/server_test.go [hunk 7]
```diff
@@ -314,23 +314,262 @@ func TestServer_PredictHandler_StandardTransformer(t *testing.T) {
...
+				statusCode: 200,
+			},
+		},
+		{
+			name:         "table transformation with feast and series transformation",
+			specYamlPath: "../pipeline/testdata/valid_feast_series_transform.yaml",
+			mockFeasts: []mockFeast{
+				{
+					request: &feastSdk.OnlineFeaturesRequest{
+						Project: "default", // used as identifier for mocking. must match config
```
Comment (pradithya) [inline api/pkg/transformer/server/server_test.go:390]: Since we don't assert the request to feast, have you make sure it's as expected?

**Comment by pradithya** [comment]
LGTM, this is awesome!!
Even better, if you can improve the assertion of request being sent to Feast (https://github.com/gojek/merlin/pull/212#discussion_r814596487)

**Review by pradithya — APPROVED** [review]

File: api/pkg/transformer/server/server_test.go (hunk 7 above)
Comment (ariefrahmansyah) [reply api/pkg/transformer/server/server_test.go:390]: Yeah, I manually checked it by logging the request here: https://github.com/gojek/merlin/pull/212/files/c9656a225bb7b0d9c3f59bac6f00d8be7c7c4686#diff-744e4800800a7662f4f3e134e1885cbd4571a9f3a802bbedeab46cf168251fe0R777

I don't like this approach though, we should be able to verify the request to the feast.

File: api/pkg/transformer/types/table/table.go [hunk 8]
```diff
@@ -367,7 +367,7 @@ func broadcastScalar(colMap map[string]interface{}, length int) map[string]inter
 				continue
 			}
 
-			val = colValueVal.Index(0)
+			val = colValueVal.Index(0).Interface()
```
Comment (tiopramayudi) [inline api/pkg/transformer/types/table/table.go:370]: why change this?

File: api/pkg/transformer/server/server_test.go [hunk 9]
```diff
@@ -482,6 +774,7 @@ func TestServer_PredictHandler_StandardTransformer(t *testing.T) {
 			for _, m := range tt.mockFeasts {
 				project := m.request.Project
 				mockFeast.On("GetOnlineFeatures", mock.Anything, mock.MatchedBy(func(req *feastSdk.OnlineFeaturesRequest) bool {
+					// log.Println("GetOnlineFeatures", req.Project, req.Entities)
```
Comment (tiopramayudi) [inline api/pkg/transformer/server/server_test.go:777]: please remove this

File: api/pkg/transformer/types/series/series.go (hunk 1 above)
Comment (tiopramayudi) [reply api/pkg/transformer/types/series/series.go:15]: ok

File: api/pkg/transformer/types/series/series_test.go [hunk 10]
```diff
@@ -102,6 +111,15 @@ func TestSeries_NewInferType(t *testing.T) {
 			want:    New([]interface{}{11, 2222, 3333, 4444, nil}, Int, "int_series"),
 			wantErr: false,
 		},
+		{
+			name: "int list",
```
Comment (tiopramayudi) [inline api/pkg/transformer/types/series/series_test.go:115]: can you add test case for mix type 
```
args: args{
				values:     [][]interface{{1111, 1.1111}, {3333, 4444}},
				seriesName: "interface_listseries",
			},
```
what is the expectation?

File: api/pkg/transformer/types/table/table.go (hunk 8 above)
Comment (ariefrahmansyah) [reply api/pkg/transformer/types/table/table.go:370]: Because, when creating series of list elements, colValueVal.Index(0) returns reflect.Value instead of interface{}. So when we pass these reflect.Value to NewInferType to create the series, it will use wrong series' type [1]->[2], because it sends reflect.Struct instead of the correct slice type.

[1] https://github.com/gojek/merlin/blob/support-list-features/api/pkg/transformer/types/series/series.go#L60-L66
[2] https://github.com/gojek/merlin/blob/support-list-features/api/pkg/transformer/types/series/series.go#L238-L260

**Comment by tiopramayudi** [comment]
lgtm

**Review by tiopramayudi — APPROVED** [review]
